### ISSUE TITLE
JWS JSON serialization?

### ISSUE THREAD

**Issue by FGasper** [DCIT/perl-Crypt-JWT#2]
Is there no support here for JWS JSON serialization?

https://tools.ietf.org/html/rfc7515#section-3.2

**Comment by karel-m** [comment]
`Crypt::JWT` is able to encode (serialize) as well as decode (deserialize) to/from JWS. 

What particularly are you missing? Or what do you need to achieve?

**Comment by FGasper** [comment]
I’ve found a way around the problem, but Crypt::JWT doesn’t appear to have the ability to encode to the JWS JSON format, which is the only way to have unprotected headers.

-FG

**Comment by karel-m** [comment]
Aaaah, now I got it.

Yes, you are right `Crypt::JWT` currently supports only JWS Compact Serialization.

**Comment by karel-m** [comment]
I can imagine a support for `Flattened JWS JSON Serialization` (== one signature only). I am afraid that `General JWS JSON Serialization` would require significant changes in module interface.

**Comment by FGasper** [comment]
On 25 Apr 2016, at 4:58 PM, karel-m notifications@github.com wrote:

> I can imagine a support for Flattened JWS JSON Serialization (== one signature only). I am afraid that General JWS JSON Serialization would require significant changes in module interface.

OK … probably good enough, then, just to document up-front that this module only does that one format.

-FG

**Comment by karel-m** [comment]
Please check/try the initial idea from https://github.com/DCIT/perl-Crypt-JWT/commit/c3a7a5211b5acd4a81963306956b2e59b2b02db8

It only handles JWS (not JWT), only Flattened (not General) JSON Serialization, no tests, no documentation.

Usage:

``` perl
use Modern::Perl;
use Crypt::JWT qw(encode_jwt decode_jwt);
use Data::Dumper;
$Data::Dumper::Terse = 1;

my $jws1 = encode_jwt(payload=>{a=>1,b=>2}, alg=>'HS256', key=>'secret');
my $jws2 = encode_jwt(payload=>{a=>1,b=>2}, alg=>'HS256', key=>'secret', serialization=>'flattened');
my $jws3 = encode_jwt(payload=>{a=>1,b=>2}, alg=>'HS256', key=>'secret', serialization=>'flattened', unprotected_headers=>{kid=>'blex',xid=>'foo'});

say "### compact   =$jws1";
say "### flattened1=$jws2";
say "### flattened2=$jws3";

my ($h1, $p1) = decode_jwt(token=>$jws1, key=>'secret', decode_header=>1);
my ($h2, $p2) = decode_jwt(token=>$jws2, key=>'secret', decode_header=>1);
my ($h3, $p3) = decode_jwt(token=>$jws3, key=>'secret', decode_header=>1);

say "### compact   :\nheader=". Dumper($h1) . "\npayload=". Dumper($p1);
say "### flattened1:\nheader=". Dumper($h2) . "\npayload=". Dumper($p2);
say "### flattened2:\nheader=". Dumper($h3) . "\npayload=". Dumper($p3);
```

**Comment by karel-m** [comment]
In the latest master I have tried to implement flattened json serialization for both JWE and JWS.

It is not tested. I have created `t/flattened.t` - if you have some spare time please try to add more tests to `t/flattened.t` or post here some examples / test vectors which can be turned into tests.

I have also updated documentation.

**Comment by FGasper** [comment]
@karel-m I’m pushing up a branch that has a test of the flattened JSON, but it doesn’t pass.

Would you be amenable to releasing with just the jwk header support, then doing flattened JSON later? Not to be selfish, but my own needs don’t encompass flattened JSON; I’m happy to continue testing it, though, but it’d be really handy if there could be a release in the next day or two.

**Comment by FGasper** [comment]
https://github.com/FGasper/perl-Crypt-JWT/tree/issue_2_test

**Comment by karel-m** [comment]
I have fixed the failing test see https://github.com/DCIT/perl-Crypt-JWT/commit/d84ce4b10b387d63d204124b0172ba0a6952dcb8

**Comment by FGasper** [comment]
Confirmed - the test passes for me now. :)

-FG

**Comment by karel-m** [comment]
I have released Crypt-JWT-0.012

Closing this issue.
